src/middleware/subscribeWithSelector.ts

```diff
@@ -1,62 +1,120 @@
-  >(
-    fn: (set: CustomSetState, get: CustomGetState, api: CustomStoreApi) => S
-  ) =>
-  (
-    set: CustomSetState,
-    get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithSubscribeWithSelector<S>
-  ): S => {
-    const origSubscribe = api.subscribe as Subscribe<S>
-    api.subscribe = ((selector: any, optListener: any, options: any) => {
-      let listener: StateListener<S> = selector // if no selector
-      if (optListener) {
-        const equalityFn = options?.equalityFn || Object.is
-        let currentSlice = selector(api.getState())
-        listener = (state) => {
-          const nextSlice = selector(state)
-          if (!equalityFn(currentSlice, nextSlice)) {
-            const previousSlice = currentSlice
-            optListener((currentSlice = nextSlice), previousSlice)
-          }
-        }
-        if (options?.fireImmediately) {
-          optListener(currentSlice, currentSlice)
+
+// ============================================================================
+// Implementation
+
+type EState = { __isState: true }
+type ESelectedState = { __isSelectedState: true }
+type EStore = Store<EState>
+type ESubscribeWithSelector = 
+  (storeInitializer: StoreInitializer<EState, EStore>) =>
+    StoreInitializer<EState, EStore & ESubscribeWithSelectorStore>
+
+interface ESubscribeWithSelectorStore
+  { subscribe:
+    ( selector: (state: EState) => ESelectedState
+    , listener:
+        ( selectedState: ESelectedState
+        , previousSelectedState: E.Previous<ESelectedState>
+        )
+          => void
+    , options?:
+        { equalityFn?: (a: ESelectedState, b: ESelectedState) => boolean
+        , fireImmediately?: boolean
         }
+    ) =>
+      () => void
+  }
+
+const subscribeWithSelectorImpl: ESubscribeWithSelector =
+  storeInitializer =>
+    (parentSet, parentGet, parentStore) => {
+
+  let parentSubscribe = parentStore.subscribe
+  let updatedParentStore = parentStore as EStore & ESubscribeWithSelectorStore
+  type UpdatedSubscribeArguments = F.O2.Arguments<(typeof updatedParentStore)['subscribe']>
+
+  updatedParentStore.subscribe = (...args: UpdatedSubscribeArguments) => {
+    if (!args[1]) {
+      pseudoAssert(args.length === 1)
+      return parentSubscribe(...args)
+    }
+  
+    pseudoAssert(args.length === 3)
+    let [selector, listener, _options] = args
+    let { equalityFn: equals, fireImmediately } =
+      { equalityFn: objectIs, fireImmediately: false, ..._options }
+  
+    let currentSelected = selector(parentGet())
+    let previousSelected = currentSelected as E.Previous<ESelectedState>
+    const emit = () => listener(currentSelected, previousSelected)
+
+    if (fireImmediately) emit()
+  
+    return parentSubscribe(() => {
+      let nextSelected = selector(parentGet())
+      if (equals(currentSelected, nextSelected)) return
+  
+      previousSelected = E.previous(currentSelected)
+      currentSelected = nextSelected
+      emit()
+    })
```

Comment:
Previously it was...
```ts
let unsubscribe = parentSubscribe(() => {
  ...
})

if (fireImmediately) emit()

return unsubscribe;
```
And now it's...
```ts
if (fireImmediately) emit()

return parentSubscribe(() => {
  ...
})
```
...so that it looks a little nicer haha, but now it changes the order of events, which I hope is okay because I don't think we guarantee that we'll only emit after subscribing.

Reply:
I would prefer the previous one. For now, the behavior is okay, but if something is changed in the future, firing after subscribing seems safer. Anyway, it's the pattern I use everywhere, so keep it consistent.

Reply:
Cool will keep it as previous